File: backend/draft-api/src/main/scala/no/ndla/draftapi/model/api/QualityEvaluationDTO.scala
/*
 * Part of NDLA draft-api
 * Copyright (C) 2024 NDLA
 *
 * See LICENSE
 *
 */

package no.ndla.draftapi.model.api

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import no.ndla.common.model.domain.draft.Grade
import sttp.tapir.Schema
import sttp.tapir.Schema.annotations.description
import no.ndla.common.DeriveHelpers

@description("Quality evaluation of the article")
case class QualityEvaluationDTO(
    @description("The grade (1-5) of the article")
    grade: Grade,
    @description("Note explaining the score")
    note: Option[String],
)

object QualityEvaluationDTO {
  implicit def encoder: Encoder[QualityEvaluationDTO] = deriveEncoder
  implicit def decoder: Decoder[QualityEvaluationDTO] = deriveDecoder
  implicit def schema: Schema[QualityEvaluationDTO]   = DeriveHelpers.getSchema
}

File: backend/taxonomy-api/src/main/kotlin/no/ndla/taxonomy/service/dtos/QualityEvaluationDTO.kt
/*
 * Part of NDLA taxonomy-api
 * Copyright (C) 2024 NDLA
 *
 * See LICENSE
 */

package no.ndla.taxonomy.service.dtos

import io.swagger.v3.oas.annotations.media.Schema
import kotlin.jvm.optionals.getOrNull
import no.ndla.taxonomy.domain.Grade
import no.ndla.taxonomy.domain.Node

@Schema
data class QualityEvaluationDTO(
    @field:Schema(description = "The grade (1-5) of the article") val grade: Grade,
    @field:Schema(description = "Note explaining the score") val note: String?,
) {
  companion object {
    fun fromNode(node: Node): QualityEvaluationDTO? =
        node.qualityEvaluationGrade.getOrNull()?.let {
          QualityEvaluationDTO(it, node.qualityEvaluationNote.getOrNull())
        }
  }
}
